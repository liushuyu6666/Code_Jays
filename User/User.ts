import mongoose, { Schema } from "mongoose";

export class User {
  public id: string;
  public username: string;
  public password: string;
  public email: string;

  constructor(
    id: string,
    username: string,
    hashedPassword: string,
    email: string
  ) {
    this.id = id;
    this.email = email;
    this.username = username;
    this.password = hashedPassword;
  }
}

const userSchema = new Schema<User>(
  {
    id: String,
    username: String,
    password: String,
    email: String,
  },
  {
    versionKey: false, // No `__v` field in mongodb
    collection: 'User'
  }
);

// Mongoose automatically looks for the plural version of your model name.
// So, the collection will be 'users', not 'user'.
// If you don't like this behavior, goto https://mongoosejs.com/docs/api/mongoose.html#example-1
export const UserModel = mongoose.model<User>("User", userSchema);
