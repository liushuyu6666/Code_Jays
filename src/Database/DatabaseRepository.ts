import bcrypt from 'bcryptjs';
import mysql, { Connection as MysqlConnection } from 'mysql';
import { MongoClient } from 'mongodb';

// TODO: configurable
const DATABASE_NAME = 'codejays';
const MONGO_URI = 'mongodb://127.0.0.1:27017/CodeJays';

export type DbCallbackFn<T> = (client: MongoClient | MysqlConnection) => Promise<T>;

export type DbOperation = <T>(callback: DbCallbackFn<T>) => Promise<T>;


export enum DatabaseType {
    MongoDB,
    MySQL
} 

export class DatabaseRepository {
    public dbOperation: DbOperation;

    constructor(databaseType: DatabaseType) {
        switch(databaseType) {
            case DatabaseType.MongoDB:
                this.dbOperation = this.mongodbOperation();
                break;
            case DatabaseType.MySQL:
                this.dbOperation = this.mysqlOperation();
                break;
            default:
                throw new Error('Invalid database type');
        }
    }

    private mongodbOperation(): DbOperation {
        return async (callback) => {
            const connection = await MongoClient.connect(MONGO_URI);
            try {
                const result = await callback(connection);
                return result;
            } catch(error) {
                throw error;
            } finally {
                await connection.close();
            }
        };
    }

    private mysqlOperation(): DbOperation {
        return async (callback) => {
            const mysqlConnection: MysqlConnection = mysql.createConnection({
                host: 'localhost',
                user: 'codejays',
                password: 'password',
                database: 'codejays',
            });
            try {
                const result = await callback(mysqlConnection);
                return result;
            } catch (error) {
                throw error;
            } finally {
                mysqlConnection.end();
            }
        }
    }

    /** For Mongodb */
    public async existsCollection(colName: string): Promise<boolean> {
        return await this.dbOperation(async (client) => {
            const col = await (client as MongoClient)
                .db()
                .listCollections({
                    name: colName,
                })
                .toArray();
            return col.length > 0;
        });
    }

    /** For Mongodb */
    public async createCollectionIfNotExists(colName: string): Promise<void> {
        const colExists = await this.existsCollection(colName);
        if (colExists) return;

        await this.dbOperation(async (client) => {
            await (client as MongoClient).db().createCollection(colName);
        });
    }

    public encryptPassword(password: string): string {
        const salt = bcrypt.genSaltSync(10);
        return bcrypt.hashSync(password, salt);
    }

    /** For Mysql */
    public async execSql(sql: string, values: (string | Date)[]): Promise<any> {
        return await this.dbOperation(async (client) => {
            return new Promise((res, rej) => {
                (client as MysqlConnection).query(sql, values, (error, result) => {
                    if (error) rej(error);
                    else res(result);
                });
            });
        });
    }

    /** For Mysql */
    public async existsTable(table: string): Promise<boolean> {
        const sql = `SELECT 1 FROM information_schema.tables WHERE table_schema = ? AND table_name = ? LIMIT 1`;
        const values = [DATABASE_NAME, table];

        const result = await this.execSql(sql, values);
        return result.length > 0;
    }

    /** For Mysql */
    public async createUserTableIfNotExists(): Promise<void> {
        const tableExists = await this.existsTable('user');
        if(tableExists) return;
        const sql = `
            CREATE TABLE user (
                userId VARCHAR(255) PRIMARY KEY,
                username VARCHAR(255) NOT NULL,
                password VARCHAR(255) NOT NULL,
                email VARCHAR(255) NOT NULL
            )
        `;
        await this.execSql(sql, []);
        
        // TODO: logger
        console.log("Create user table in mysql");
    }

    /** For Mysql */
    public async createImageTableIfNotExists(): Promise<void> {
        const tableExists = await this.existsTable('image');
        if (tableExists) return;
        const sql = `
            CREATE TABLE image (
                imageId VARCHAR(255) PRIMARY KEY,
                userId VARCHAR(255) NOT NULL,
                fileName VARCHAR(255) NOT NULL,
                url VARCHAR(255) NOT NULL,
                uploadDate TIMESTAMP NOT NULL
            )
        `;
        await this.execSql(sql, []);

        // TODO: logger
        console.log('Create image table in mysql');
    }
}