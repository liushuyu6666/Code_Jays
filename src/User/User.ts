import mongoose, { Schema } from "mongoose";

export class User {
  public userId: string;
  public username: string;
  public password: string;
  public email: string;

  constructor(
    userId: string,
    username: string,
    hashedPassword: string,
    email: string
  ) {
    this.userId = userId;
    this.email = email;
    this.username = username;
    this.password = hashedPassword;
  }
}

const userSchema = new Schema<User>(
  {
    userId: String,
    username: String,
    password: String,
    email: String,
  },
  {
    versionKey: false, // No `__v` field in mongodb
    collection: 'User'
  }
);

// Mongoose automatically looks for the plural version of your model name.
// So, the collection will be 'users', not 'user'.
// If you don't like this behavior, goto https://mongoosejs.com/docs/api/mongoose.html#example-1
export const UserModel = mongoose.model<User>("User", userSchema);
