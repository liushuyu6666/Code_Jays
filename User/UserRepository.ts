import { v4 as uuidv4 } from "uuid";
import bcrypt from 'bcryptjs';

import { User, UserModel } from './User';

export class UserRepository {
    static async addUser(username: string, password: string, email: string): Promise<User> {
        // Encrypt the password
        const salt = bcrypt.genSaltSync(10);
        const hashedPassword = bcrypt.hashSync(password, salt);

        // Generate id
        const id = uuidv4();

        const user = new User(id, username, hashedPassword, email);

        const newUser = new UserModel(user);
        await newUser.save();
        return newUser;

        // return await UserModel.create(newUser);
    }

    static async getUserByEmail(email: string): Promise<User | null> {
        return UserModel.findOne({ email }).exec();
    }
}
