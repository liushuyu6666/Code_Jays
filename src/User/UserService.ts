import { DbOperation } from '../../db';
import { User } from './User';
import {
    DatabaseType,
    UserRepository,
    UserRepositoryFactory,
} from './UserRepository';
import bcrypt from 'bcryptjs';
import { Connection as MysqlConnection } from 'mysql';

export class UserService {
    private userRepository: UserRepository;

    constructor(databaseType: DatabaseType, dbOperation: DbOperation) {
        this.userRepository =
            UserRepositoryFactory.createUserRepository(databaseType, dbOperation);
    }

    async registerUser(
        username: string,
        password: string,
        email: string,
    ): Promise<User | string> {
        // Check if user with the same email already exists.
        const existingUser = await this.userRepository.getUserByEmail(email);

        if (existingUser) {
            return 'User with the same email already exists';
        }

        // Create a new User.
        // TODO: handle with the error from database.
        const newUser = this.userRepository.addUser(username, password, email);

        return newUser;
    }

    async verifyUserAndReturnUserId(
        email?: string,
        password?: string,
    ): Promise<string | undefined> {
        if (!email) {
            return undefined;
        }

        const loginUser = await this.userRepository.getUserByEmail(email);

        if (!loginUser || !password) {
            return undefined;
        }

        if (bcrypt.compareSync(password, loginUser?.password)) {
            return loginUser.userId;
        } else {
            return undefined;
        }
    }
}
