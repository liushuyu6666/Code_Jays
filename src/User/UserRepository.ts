import { v4 as uuidv4 } from 'uuid';
import { User } from './User';
import { MongoClient } from 'mongodb';
import { DatabaseRepository, DatabaseType } from '../Database/DatabaseRepository';

// TODO: configurable
// const DATABASE_NAME = 'codejays';

export interface UserRepository {
    addUser(username: string, password: string, email: string): Promise<User>;
    getUserByEmail(email: string): Promise<User | undefined>;
}

export class MongodbUserRepository extends DatabaseRepository implements UserRepository {
    constructor()  {
        super(DatabaseType.MongoDB);
    }
    
    async addUser(
        username: string,
        password: string,
        email: string,
    ): Promise<User> {
        await this.createCollectionIfNotExists('User');
    
        const hashedPassword = super.encryptPassword(password);

        // Generate id
        const id = uuidv4();

        await this.dbOperation(async (client) => {
            const col = (client as MongoClient).db().collection('User');
            return await col.insertOne({
                username,
                password: hashedPassword,
                email,
                userId: id
            });
        })

        return new User(id, username, hashedPassword, email);
    }

    async getUserByEmail(email: string): Promise<User | undefined> {
        if(!await this.existsCollection('User')) {
            return undefined;
        }

        const user = await this.dbOperation(async (client) => {
            const col = (client as MongoClient).db().collection('User');
            return await col.findOne({ email });
        });

        if(!user) return undefined;
        return new User(user.userId, user.username, user.password, user.email);
    }
}

export class MysqlUserRepository extends DatabaseRepository implements UserRepository {
    constructor()  {
        super(DatabaseType.MySQL);
    }

    async addUser(
        username: string,
        password: string,
        email: string,
    ): Promise<User> {
        await this.createUserTableIfNotExists();

        // Encrypt the password
        const hashedPassword = super.encryptPassword(password);

        // Generate id
        const id = uuidv4();

        const sql =
            'INSERT INTO user (userId, username, email, password) VALUES (?, ?, ?, ?)';
        const values = [id, username, email, hashedPassword];

        const result = await this.execSql(sql, values);

        return new User(id, username, hashedPassword, email);
    }

    async getUserByEmail(email: string): Promise<User | undefined> {
        if(!await this.existsTable('user')) {
            return undefined;
        }
    
        const sql = 'SELECT * FROM user WHERE email = ? LIMIT 1';
        const values = [email];

        const result = await this.execSql(sql, values);

        if(result && result.length > 0) {
            const userInfo = result[0];
            return new User(userInfo.userId, userInfo.username, userInfo.password, userInfo.email);
        } else {
            return undefined;
        }
    }
}

export class UserRepositoryFactory {
    static createUserRepository(databaseType: DatabaseType): UserRepository {
        switch(databaseType) {
            case DatabaseType.MongoDB:
                return new MongodbUserRepository();
            case DatabaseType.MySQL:
                return new MysqlUserRepository();
            default:
                throw new Error('Invalid database type');
        }
    }
}
