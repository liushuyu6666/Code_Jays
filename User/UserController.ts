import { Request, Response } from "express";
import { User } from "./User";
import { UserService } from "./UserService";
import { AuthService } from "../Auth/AuthService";
import * as httpContext from 'express-http-context';


const users: User[] = []; // In-memory storage for registered users

export class UserController {
    static async registerUser(req: Request, res: Response): Promise<Response<any, Record<string, any>>> {
        console.log('register');
        const { username, password, email } = req.body;

        console.log(httpContext);
    
        // Validate input
        if (!username || !password || !email) {
            return res
                .status(400)
                .json({ error: "Username, password, and email are required" });
        }
    
        const newUserOrError = await UserService.registerUser(username, password, email);
        if(typeof newUserOrError === 'string') {
            return res.status(409).json({ error: newUserOrError });
        }
    
        const id = newUserOrError.id;
        return res.status(201).json({ message: `User registered successfully`});
    };

    static async loginUser(req: Request, res: Response): Promise<Response<any, Record<string, any>>> {
        const { password, email, id } = req.body;

        const loginSuccess = await UserService.verifyUser(email, password);
        if(!loginSuccess) {
            return res.status(401).json({ error: 'Invalid email or password' });
        }

        // TODO: add logger to display the login user
        const token = AuthService.generateToken(id);
        return res.status(200).json({ message: 'User logged in successfully', token });
    }
}
