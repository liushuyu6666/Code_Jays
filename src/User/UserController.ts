import { Request, Response } from "express";
import { UserService } from "./UserService";
import { AuthService } from "../Auth/AuthService";
import { Connection as MysqlConnection } from 'mysql';
import { DatabaseType } from "./UserRepository";
import { DbOperation } from "../../db";

export class UserController {
    private userService: UserService;

    constructor(databaseType: DatabaseType, dbOperation: DbOperation) {
        this.userService = new UserService(databaseType, dbOperation);
    }

    async registerUser(req: Request, res: Response): Promise<Response<any, Record<string, any>>> {
        const { username, password, email } = req.body;
    
        // Validate input
        if (!username || !password || !email) {
            return res
                .status(400)
                .json({ error: "Username, password, and email are required" });
        }
    
        const newUserOrError = await this.userService.registerUser(username, password, email);
        if(typeof newUserOrError === 'string') {
            return res.status(409).json({ error: newUserOrError });
        }
    
        const id = newUserOrError.userId;
        return res.status(201).json({ message: `User registered successfully`});
    };

    async loginUser(req: Request, res: Response): Promise<Response<any, Record<string, any>>> {
        const { password, email } = req.body;

        const userId = await this.userService.verifyUserAndReturnUserId(email, password);
        if(!userId) {
            return res.status(401).json({ error: 'Invalid email or password' });
        }

        // TODO: add logger to display the login user
        const token = AuthService.generateToken(userId);
        return res.status(200).json({ message: 'User logged in successfully', token });
    }
}
