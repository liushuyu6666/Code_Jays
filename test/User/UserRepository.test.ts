import { MongodbUserRepository, MysqlUserRepository, UserRepositoryFactory } from '../../src/User/UserRepository';
import { DatabaseRepository, DatabaseType } from '../../src/Database/DatabaseRepository';
import { MongoClient } from 'mongodb';

jest.spyOn(DatabaseRepository.prototype, 'encryptPassword').mockReturnValue(
    'hashedPassword',
);

jest.mock('uuid', () => ({
    v4: jest.fn().mockReturnValue('mocked-uuid'),
}));

const mockMongoClient = {
    close: jest.fn().mockResolvedValue({}),
    db: jest.fn().mockReturnValue({
        collection: jest.fn().mockReturnValue({
            insertOne: jest.fn().mockResolvedValue(undefined),
            findOne: jest.fn(), // Return value of the method can be modified upon different cases
        }),
    }),
} as any;

jest.spyOn(MongoClient.prototype, 'connect').mockImplementation(
    () => mockMongoClient,
);

describe('MongodbUserRepository', () => {
    let mongodbUserRepository: MongodbUserRepository;

    beforeEach(() => {
        mongodbUserRepository = new MongodbUserRepository();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('addUser', () => {
        test('should return the User', async () => {
            jest.spyOn(mongodbUserRepository, 'createCollectionIfNotExists').mockResolvedValue(undefined);

            const result = await mongodbUserRepository.addUser(
                'username',
                'password',
                'email',
            );

            expect(result).toMatchSnapshot();
        });
    });

    describe('getUserByEmail', () => {
        test('should return undefined if no such collection', async () => {
            jest.spyOn(mongodbUserRepository, 'existsCollection').mockResolvedValue(false);

            const result = await mongodbUserRepository.getUserByEmail('email');

            expect(result).toBe(undefined);
        });

        test('should return undefined if no such user', async () => {
            jest.spyOn(mongodbUserRepository, 'existsCollection').mockResolvedValue(true);

            mockMongoClient.db().collection().findOne = jest.fn().mockResolvedValue(undefined)

            const result = await mongodbUserRepository.getUserByEmail('email');

            expect(result).toBe(undefined);
        });

        test('should return the user if it exists', async () => {
            jest.spyOn(mongodbUserRepository, 'existsCollection').mockResolvedValue(true);

            mockMongoClient.db().collection().findOne = jest.fn().mockResolvedValue({
                userId: 'userId2',
                username: 'username2',
                hashedPassword: 'hashedPassword2',
                email: 'email2',
            })

            const result = await mongodbUserRepository.getUserByEmail('email2');

            expect(result).toMatchSnapshot();
        })
    })
});

describe('MysqlUserRepository', () => {
    let mysqlUserRepository: MysqlUserRepository;

    beforeEach(() => {
        mysqlUserRepository = new MysqlUserRepository();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('addUser', () => {
        test('should return the inserted user', async () => {
            jest.spyOn(mysqlUserRepository, 'createUserTableIfNotExists').mockResolvedValue(undefined);
            jest.spyOn(mysqlUserRepository, 'execSql').mockResolvedValue(undefined);

            const result = await mysqlUserRepository.addUser(
                'username_mysql',
                'password_mysql',
                'email@mysql',
            );

            expect(result).toMatchSnapshot();
        });
    });

    describe('getUserByEmail', () => {
        test('should return undefined if no such table', async () => {
            jest.spyOn(mysqlUserRepository, 'existsTable').mockResolvedValue(false);

            const result = await mysqlUserRepository.getUserByEmail('email');

            expect(result).toBe(undefined);
        });

        test('should return the user if it exists', async () => {
            jest.spyOn(mysqlUserRepository, 'existsTable').mockResolvedValue(true);
            jest.spyOn(mysqlUserRepository, 'execSql').mockResolvedValue([{
                userId: 'userId2_mysql',
                username: 'username2_mysql',
                hashedPassword: 'hashedPassword2_mysql',
                email: 'email2@mysql',
            }]);

            const result = await mysqlUserRepository.getUserByEmail('email2@mysql');

            expect(result).toMatchSnapshot();
        });

        test('should return undefined if the user does not exist', async () => {
            jest.spyOn(mysqlUserRepository, 'existsTable').mockResolvedValue(true);
            jest.spyOn(mysqlUserRepository, 'execSql').mockResolvedValue([]);

            const result = await mysqlUserRepository.getUserByEmail('email2@mysql');

            expect(result).toBe(undefined);
        });
    })
});

describe('Test UserRepositoryFactory', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should return MongodbUserRepository if it is MongoDB', () => {
        const result = UserRepositoryFactory.createUserRepository(DatabaseType.MongoDB);
        expect(result).toBeInstanceOf(MongodbUserRepository);
    });

    test('should return MysqlImageRepository if it is Mysql', () => {
        const result = UserRepositoryFactory.createUserRepository(DatabaseType.MySQL);
        expect(result).toBeInstanceOf(MysqlUserRepository);
    });

    test('should throw an error for an invalid database type', () => {
        expect(() => {
            UserRepositoryFactory.createUserRepository('InvalidDatabaseType' as unknown as  DatabaseType);
          }).toThrowError('Invalid database type');
    });
})
