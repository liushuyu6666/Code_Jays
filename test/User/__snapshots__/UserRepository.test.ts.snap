// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongodbUserRepository addUser should return the User 1`] = `
User {
  "email": "email",
  "password": "hashedPassword",
  "userId": "mocked-uuid",
  "username": "username",
}
`;

exports[`MongodbUserRepository getUserByEmail should return the user if it exists 1`] = `
User {
  "email": "email2",
  "password": undefined,
  "userId": "userId2",
  "username": "username2",
}
`;

exports[`MysqlUserRepository addUser should return the inserted user 1`] = `
User {
  "email": "email@mysql",
  "password": "hashedPassword",
  "userId": "mocked-uuid",
  "username": "username_mysql",
}
`;

exports[`MysqlUserRepository getUserByEmail should return the user if it exists 1`] = `
User {
  "email": "email2@mysql",
  "password": undefined,
  "userId": "userId2_mysql",
  "username": "username2_mysql",
}
`;
